runbotics/runbotics-desktop/src/action/google/google.action-handler.ts

```diff
@@ -1,63 +1,217 @@
+    GoogleSheetGetCellActionInput,
+    CellByValueActionOutput,
+} from './google.types';
+import { GoogleAction } from 'runbotics-common';
+import {
+    getCellsByValueInputSchema,
+    getCellInputSchema,
+    getCellsInputSchema,
+    getWorksheetInputSchema,
+    setCellsInputSchema,
+} from './google.utils';
+import { ZodError } from 'zod';
+import { formatZodError } from '#utils/zodError';
 
 @Injectable()
 export default class GoogleActionHandler extends StatelessActionHandler {
-    constructor(
-        private readonly serverConfigService: ServerConfigService,
-    ) {
+    private readonly logger = new RunboticsLogger(GoogleActionHandler.name);
+
+    constructor(private readonly serverConfigService: ServerConfigService) {
         super();
     }
 
-    async write(input: GoogleSheetWriteActionInput): Promise<GoogleSheetWriteActionOutput> {
-        const tokens = {
-            refresh_token: this.serverConfigService.googleRefreshToken,
-            scope: 'https://www.googleapis.com/auth/drive https://www.googleapis.com/auth/spreadsheets',
-            token_type: 'Bearer',
-        };
+    async getWorksheet(rawInput: GoogleSheetGetWorksheetActionInput) {
+        const { spreadsheetId, sheet } = await getWorksheetInputSchema
+            .parseAsync(rawInput)
+            .catch((error: ZodError) => {
+                throw formatZodError(error);
+            });
 
-        const oauth2Client = new google.auth.OAuth2(
-            '373359756354-7bdu5c7ibjoiprlak9snb8rgpahnq06q.apps.googleusercontent.com',
-            'KAR6T2_qDA7c-co_pdbwLjRb',
-            'http://localhost:3333',
-        );
-        await oauth2Client.setCredentials(tokens);
-        const sheets = new Sheets({ auth: oauth2Client });
+        const sheets = this.getGoogleSheets();
 
-        const spreadsheet = await sheets.spreadsheets.get({
-            spreadsheetId: '13-INLpquqKIBtMrHFAZWCRpWkLcthyMqDkMlGpdv6U8',
+        const defaultSheet = await this.getDefaultSheet(spreadsheetId);
+
+        const { data: sheetData } = await sheets.spreadsheets.values.get({
+            range: sheet ? sheet.split('!')[0] : defaultSheet,
+            spreadsheetId,
         });
 
-        const body: Schema$ValueRange = {
-            values: input.values,
-        };
-        await sheets.spreadsheets.values.update({
-            spreadsheetId: spreadsheet.data.spreadsheetId!,
-            range: input.range ? input.range : 'A1',
+        return this.mapValueRange(sheetData);
+    }
+
+    async getCell(rawInput: GoogleSheetGetCellActionInput) {
+        const { spreadsheetId, cell, sheet } = await getCellInputSchema
+            .parseAsync(rawInput)
+            .catch((error: ZodError) => {
+                throw formatZodError(error);
+            });
+
+        const sheets = this.getGoogleSheets();
+
+        const defaultSheet = await this.getDefaultSheet(spreadsheetId);
+
+        const { data: sheetData } = await sheets.spreadsheets.values.get({
```

Comment:
Nice check. Since this ! was not quite clear for me (that sheet can be also passed as **sheetName!range** value) I would add link to documentation which point to this - can be useful for the feature reference.